### PR TITLE
Bump rustc nightly-2022-09-20

### PR DESCRIPTION
**Pull request by MichaelMauderer**
### Pull Request Description

Bump rustc nightly-2022-08-30 and fix new errors and lints. 

### Important Notes

* removes `DEPRECATED_Tween` from the animation framework, which failed to compile with a newer version of rustc. All uses were replaced with the newer `Easing` instead.
* Removed some potential UB indicated by Clippy in `lib/rust/data-structures/src/diet.rs` @mwu-tow, please double-check this. 

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: lib/rust/automata/src/dfa.rs [hunk 1]
```diff
@@ -71,13 +71,15 @@ impl Dfa {
...
+
         let mut out = String::new();
         for row in 0..self.links.rows {
-            out += &format!("node_{}[label=\"{}\"]\n", row, row);
+            writeln!(out, "node_{}[label=\"{}\"]", row, row).unwrap();
             for column in 0..self.links.columns {
                 let state = self.links[(row, column)];
                 if !state.is_invalid() {
-                    out += &format!("node_{} -> node_{}\n", row, state.id());
+                    writeln!(out, "node_{} -> node_{}", row, state.id()).unwrap();
```
Comment (wdanilo) [inline lib/rust/automata/src/dfa.rs:82]: I don't like these changes TBH. They are using `unwrap` now and no unwrap in the previous version. Why these changes?

File: build/enso-formatter/src/main.rs [hunk 2]
```diff
@@ -284,8 +286,10 @@ fn print_h2(
     tokens: &[HeaderToken],
     str: &str,
 ) {
+    use std::fmt::Write;
+
     if tokens.iter().map(|tok| map.contains_key(tok)).any(|t| t) {
-        out.push_str(&format!("// === {} ===\n", str));
+        writeln!(out, "// === {} ===", str).unwrap()
```
Comment (wdanilo) [inline build/enso-formatter/src/main.rs:292]: I don't like these changes TBH. They are using `unwrap` now and no unwrap in the previous version. Why these changes?

File: build/enso-formatter/src/main.rs [hunk 3]
```diff
@@ -268,12 +268,14 @@ fn print_h1(
...
-        out.push('\n');
-        out.push_str(&format!("// ===={}====\n", "=".repeat(str.len())));
-        out.push_str(&format!("// === {} ===\n", str));
-        out.push_str(&format!("// ===={}====\n", "=".repeat(str.len())));
-        out.push('\n');
+        writeln!(out).unwrap();
+        writeln!(out, "// ===={}====", "=".repeat(str.len())).unwrap();
+        writeln!(out, "// === {} ===", str).unwrap();
+        writeln!(out, "// ===={}====", "=".repeat(str.len())).unwrap();
+        writeln!(out).unwrap();
```
Comment (wdanilo) [inline build/enso-formatter/src/main.rs:278]: I don't like these changes TBH. They are using `unwrap` now and no unwrap in the previous version. Why these changes?

File: lib/rust/automata/src/nfa.rs [hunk 4]
```diff
@@ -243,21 +243,25 @@ impl Nfa {
...
                     link.target.id(),
                     link.display_symbols()
-                );
+                )
+                .unwrap();
             }
             for link in &state.epsilon_links {
-                out += &format!("node_{} -> node_{}[style=dashed]\n", ix, link.id());
+                writeln!(out, "node_{} -> node_{}[style=dashed]", ix, link.id()).unwrap();
             }
```
Comment (wdanilo) [inline lib/rust/automata/src/nfa.rs:265]: I don't like these changes TBH. They are using `unwrap` now and no unwrap in the previous version. Why these changes?

File: lib/rust/config-reader/src/lib.rs [hunk 5]
```diff
@@ -93,12 +95,14 @@ pub fn generate_config_module_from_yaml(config_path: impl AsRef<std::path::Path>
...
+                writeln!(def, "{}pub {}: &'static str,", indent, key).unwrap();
+                writeln!(inst, "{}{}: \"{}\",", indent, key, value).unwrap();
+                writeln!(
+                    vars,
+                    "#[allow(non_upper_case_globals)]\npub const {}: &str = \"{}\";",
                     key, value
-                ));
+                )
+                .unwrap();
             },
```
Comment (wdanilo) [inline lib/rust/config-reader/src/lib.rs:106]: I don't like these changes TBH. They are using `unwrap` now and no unwrap in the previous version. Why these changes?

File: lib/rust/data-structures/src/diet.rs [hunk 6]
```diff
@@ -143,12 +143,9 @@ pub struct $name {
...
 
-    /// Create an empty data array. This function is safe because the intervals are build out of
-    /// [`uint`]s, which can be initialized from raw memory. Follow the link to learn more:
-    /// https://doc.rust-lang.org/std/mem/union.MaybeUninit.html.
-    #[allow(unsafe_code)]
+    /// Create an empty data array.
     pub (crate) fn empty_data_array() -> DataArray {
-        unsafe { MaybeUninit::uninit().assume_init() }
+        [default();$num]
     }
```
Comment (wdanilo) [inline lib/rust/data-structures/src/diet.rs:149]: Why this change? You've written that you removed a potential UB. Could you describe an example situation why this UB can happen and why the docs that were here are invalid? They were describing that it is safe – when it could not be safe.

File: lib/rust/ensogl/component/text/embedded-fonts/build.rs [hunk 7]
```diff
@@ -0,0 +1 @@
+
```
Comment (wdanilo) [inline lib/rust/ensogl/component/text/embedded-fonts/build.rs:1]: Is this an empty file?

File: lib/rust/ensogl/core/src/display/style/sheet.rs [hunk 8]
```diff
@@ -135,6 +135,7 @@ impl SheetNode {
 #[allow(missing_docs)]
 pub struct Expression {
     pub args:     Vec<Path>,
+    #[allow(clippy::type_complexity)]
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/style/sheet.rs:138]: is it really needed here? This type does not look so complex. If clippy warns here, maybe ithere is a setting for it in clippy conf now - is there one?

File: lib/rust/ensogl/core/src/display/style/sheet.rs [hunk 9]
```diff
@@ -191,11 +192,13 @@ impls! { From<&StaticPath> for Expression { |t| Self::reference(t) }}
 #[derive(Clone)]
 pub struct BoundExpression {
     args:     Vec<Index<Query>>,
+    #[allow(clippy::type_complexity)]
     function: Rc<dyn Fn(&[&Data]) -> Data>,
 }
 
 impl BoundExpression {
     /// Constructor.
+    #[allow(clippy::type_complexity)]
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/style/sheet.rs:201]: is it really needed here? This type does not look so complex. If clippy warns here, maybe ithere is a setting for it in clippy conf now - is there one?

File: lib/rust/ensogl/core/src/gui/cursor.rs [hunk 10]
```diff
@@ -303,8 +303,8 @@ impl Cursor {
             });
 
             eval frp.set_style([host_attached_weight,size,offset,model] (new_style) {
-                host_attached_weight.stop_and_rewind();
-                if new_style.host.is_some() { host_attached_weight.start() }
+                host_attached_weight.stop_and_rewind(0.0);
+                if new_style.host.is_some() { host_attached_weight.target(1.0) }
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/gui/cursor.rs:307]: just double checking - are you 100% sure it works the exactly same way as the previous version?

File: lib/rust/ensogl/core/src/gui/style.rs [hunk 11]
```diff
@@ -72,6 +72,7 @@ macro_rules! define_style {( $( $(#$meta:tt)* $field:ident : $field_type:ty),* $
     /// that cursor does not implement any complex style management (like pushing or popping a style
     /// from a style stack) on purpose, as it is stateful, while it is straightforward to implement
     /// it in FRP.
+    #[allow(clippy::derive_partial_eq_without_eq)]
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/gui/style.rs:75]: no comment why we dont want Eq here. Please add explanation.

File: lib/rust/prelude/src/test/stream.rs [hunk 12]
```diff
@@ -83,6 +83,7 @@ pub trait StreamTestExt<S: ?Sized + Stream> {
     ///
     /// Takes a list of predicates. Items are matched against them, after predicate succeeds match
     /// it is removed from the list.
+    #[allow(clippy::type_complexity)]
```
Comment (wdanilo) [inline lib/rust/prelude/src/test/stream.rs:86]: The same comment as above - this type doesnt look so complex.

File: lib/rust/profiler/data/src/bin/devtools.rs [hunk 13]
```diff
@@ -133,7 +133,7 @@ impl<'p, Metadata> IntervalTranslator<'p, Metadata> {
             self.events.push(event);
         }
         for child in &active.children {
-            self.visit_interval(*child, row + 1);
+            self.visit_interval(*child, _row + 1);
```
Comment (wdanilo) [inline lib/rust/profiler/data/src/bin/devtools.rs:136]: why this change? `_row` seems to be used.

**Review by wdanilo — CHANGES_REQUESTED** [review]

File: lib/rust/automata/src/dfa.rs (hunk 1 above)
Comment (MichaelMauderer) [reply lib/rust/automata/src/dfa.rs:82]: This follows from this Clippy lint: https://rust-lang.github.io/rust-clippy/master/#format_push_string
(same as the other similar changes)

Short summary: this should be a performance improvement as it avoids an additional heap allocation that happens with the `format` macro. The `unwrap` never fails.

File: lib/rust/data-structures/src/diet.rs (hunk 6 above)
Comment (MichaelMauderer) [reply lib/rust/data-structures/src/diet.rs:149]: According to the link provided, any integer initialization with uninitialised memory is UB, directly contradicting the removed comment. 

The example from the linked site, which is equivalent to what is happening here, is:
`let x: i32 = unsafe { MaybeUninit::uninit().assume_init() }; // undefined behavior! ⚠️`

No distinction between signed/unsigned integer is made in the docs: 

> (Notice that the rules around uninitialized integers are not finalized yet, but until they are, it is advisable to avoid them.)

File: lib/rust/ensogl/component/text/embedded-fonts/build.rs (hunk 7 above)
Comment (MichaelMauderer) [reply lib/rust/ensogl/component/text/embedded-fonts/build.rs:1]: This seems to be a rebase issue that I missed. Needs to be removed.

File: lib/rust/ensogl/core/src/display/style/sheet.rs (hunk 9 above)
Comment (MichaelMauderer) [reply lib/rust/ensogl/core/src/display/style/sheet.rs:201]: There is a configuration parameter `type-complexity-threshold`, which we could raise.

File: lib/rust/ensogl/core/src/gui/cursor.rs (hunk 10 above)
Comment (MichaelMauderer) [reply lib/rust/ensogl/core/src/gui/cursor.rs:307]: I made this change a while ago, but I remember this being functionally equivalent. We should carefully test that this has no corner cases, but otherwise, yes this should be fine.

File: lib/rust/ensogl/core/src/gui/style.rs (hunk 11 above)
Comment (MichaelMauderer) [reply lib/rust/ensogl/core/src/gui/style.rs:75]: The reason is, that we don't know what types this struct will be instantiated with. So, sometimes we might not be able to derive `Eq` because of floats, but other structs might not use floats, and will then be flagged by clippy.

File: lib/rust/profiler/data/src/bin/devtools.rs (hunk 13 above)
Comment (MichaelMauderer) [reply lib/rust/profiler/data/src/bin/devtools.rs:136]: Clippy complained because the variable is used only in recursion. Apparently, variables only used for recursion should have a leading underscore.

Comment (MichaelMauderer) [reply lib/rust/profiler/data/src/bin/devtools.rs:136]: Blergh. I think I missed what cluppy was trying to tell me: that variable is ONLY used in recursion. It seems not to do anything and can be removed. (On my phone, so I need to double check this, but either cluppy is right or we have a false positive and should ignore this lint)

File: lib/rust/automata/src/dfa.rs (hunk 1 above)
Comment (wdanilo) [reply lib/rust/automata/src/dfa.rs:82]: Performance improvement is a clear win here, however, this solution has downsides, just as the link that you pasted describes:

image

I'm OK with using `writeln!` macro, but I'm not ok with unwraps. Sure, they do nothing here, but unwraps make code smelly. If you want to tell that the result's result is ok and we dont care about it, use `.ok()` instead.

File: lib/rust/data-structures/src/diet.rs (hunk 6 above)
Comment (wdanilo) [reply lib/rust/data-structures/src/diet.rs:149]: You are right, the docs are not correct. However, I'm not sure whether this data structure ever tries to access elements that were not initialized by some logic in this structure. Initializing all elements upfront can cause performance drop, and this is a low-level struct. As we are not using it anywhere in our codebase (we are just referencing it from the docs), I'd propose to chose the first solution described below. I see 2 possible solutions here:

1. Leaving it as it was + fixme about wrong docs + potential UB to be checked one day.
2. Fixing it properly - this includes analysis if these elements really need to be initialized, benchmarking vs the old version, and fixing docs.

File: lib/rust/ensogl/component/text/embedded-fonts/build.rs (hunk 7 above)
Comment (wdanilo) [reply lib/rust/ensogl/component/text/embedded-fonts/build.rs:1]: Cool, so why is it not removed yet? :)

File: lib/rust/ensogl/core/src/display/style/sheet.rs (hunk 9 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/style/sheet.rs:201]: What is the status of this? + What is the threshold that makes so simple types being pointed out?

File: lib/rust/data-structures/src/diet.rs (hunk 6 above)
Comment (MichaelMauderer) [reply lib/rust/data-structures/src/diet.rs:149]: Note that the UB is not caused by *accessing* but merely by initialising. The correct solution would have to propagate the `MaybeUninit` until after initialisation. I don't think this should be part of this PR, instead we should file a new task to investigate this. We are hotfixing UB here, aplying a new performance improvement is somewhat less urgent.

File: lib/rust/profiler/data/src/bin/devtools.rs (hunk 13 above)
Comment (mwu-tow) [reply lib/rust/profiler/data/src/bin/devtools.rs:136]: Removed.

File: lib/rust/ensogl/core/src/gui/style.rs (hunk 11 above)
Comment (mwu-tow) [reply lib/rust/ensogl/core/src/gui/style.rs:75]: Added comment.

File: lib/rust/ensogl/core/src/display/style/sheet.rs (hunk 9 above)
Comment (mwu-tow) [reply lib/rust/ensogl/core/src/display/style/sheet.rs:201]: Fixed on my branch. The default is `250`.

File: lib/rust/ensogl/component/text/embedded-fonts/build.rs (hunk 7 above)
Comment (mwu-tow) [reply lib/rust/ensogl/component/text/embedded-fonts/build.rs:1]: Just removed it in my branch.

**Comment by mwu-tow** [comment]
Superseded by #3743.
